elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/AutoGrowingPagedBuffer.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.elasticsearch.util.impl;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.CharacterCodingException;
+import java.nio.charset.CharsetEncoder;
+import java.nio.charset.CoderResult;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.LinkedList;
+
+import org.apache.http.nio.ContentEncoder;
+
+/**
+ * An automatically growing, paged buffer.
+ * <p>
+ * To be used when you need an infinite-size buffer,
+ * for example when your class is pushed content to
+ * and generate bytes from this content,
+ * but cannot always flush them to your consumer.
+ *
+ * @author Yoann Rodiere
+ */
+class AutoGrowingPagedBuffer {
```

Comment:
I like the design idea behind this class, but I'm wondering if you're not just merely re-writing the inner class I had created within `GsonHttpEntity` into a separate file.

When one class exists purely to simplify another one, and every single line of its design is influenced by this other class, there's a very high degree of coupling so I generally think it should be an inner class of `GsonHttpEntity`.

Thinking about your design process: I feel like you creates this in a new file to have a clean slate; that's understandable as you were taking ownership of the twisted stuff I had created before :)
Still, don't you think it belongs to be moved back in there now that you've fleshed it out?

Reply:
I agree that `ProgressiveCharBuffer` heavily relies on `AutoGrowingPageBuffer`. I agree we could merge `ProgressiveCharBuffer` with `AutoGrowingPageBuffer`.
But I'm not sure I agree there is a high degree of coupling between `GsonHttpEntity` and `AutoGrowingPageBuffer`. `ProgressiveCharBuffer` does not depend on `GsonHttpEntity` at all. And theoretically, we could perfectly well use `ProgressiveCharBuffer` for another type of entity.
Also, there's the matter of code being easily understood: nested classes are nice and all, but when files grow in size it becomes harder to understand to full extent of what they do (at least it's more difficult for me).

I'll merge `AutoGrowingPageBuffer` with `ProgressiveCharBufferWriter`, but leave them in a separate file. Would that be ok to you?

Reply:
Yes sounds good. I agree that excessively large classes need to be split.